man/man1/proj.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "PROJ" "1" "15 Sep 2026" "9.9" "PROJ"
.SH NAME
proj \- Cartographic projection filter
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
\fBproj\fP [\fB\-beEfiIlmorsStTvVwW\fP] [args]] ([\fI+opt[=arg]\fP ...] | {crs}) file ...
.sp
\fBinvproj\fP [\fB\-beEfiIlmorsStTvVwW\fP] [args]] ([\fI+opt[=arg]\fP ...] | {crs}) file ...
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
\fI\%proj\fP and \fI\%invproj\fP perform respective forward and inverse
conversion of cartographic data to or from cartesian data with a wide
range of selectable projection functions.
.sp
\fI\%invproj\fP may not be available on all platforms; in this case
use \fI\%proj \-I\fP instead.
.sp
The following control parameters can appear in any order
.INDENT 0.0
.TP
.B \-b
Special option for binary coordinate data input and output through standard
input and standard output. Data is assumed to be in system type double
floating point words. This option is to be used when \fI\%proj\fP is a child process
and allows bypassing formatting operations.
.UNINDENT
.INDENT 0.0
.TP
.B \-d <n>
.UNINDENT
.sp
Added in version 5.2.0: Specify the number of decimals to round to in the output.

.INDENT 0.0
.TP
.B \-i
Selects binary input only (see \fI\%\-b\fP).
.UNINDENT
.INDENT 0.0
.TP
.B \-I
Alternate method to specify inverse projection. Redundant when used with
\fI\%invproj\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-o
Selects binary output only (see \fI\%\-b\fP).
.UNINDENT
.INDENT 0.0
.TP
.B \-t<a>
Where \fIa\fP specifies a character employed as the first character to denote a
control line to be passed through without processing. This option
applicable to ASCII input only. (# is the default value).
.UNINDENT
.INDENT 0.0
.TP
.B \-e <string>
Where \fIstring\fP is an arbitrary string to be output if an error is detected during
data transformations. The default value is a three character string: \fB*\et*\fP\&.
Note that if the \fI\%\-b\fP, \fI\%\-i\fP or \fI\%\-o\fP options are employed, an error
is returned as HUGE_VAL value for both return values.
.UNINDENT
.INDENT 0.0
.TP
.B \-E
Causes the input coordinates to be copied to the output line prior to
printing the converted values.
.UNINDENT
.INDENT 0.0
.TP
.B \-l<[=id]>
List projection identifiers that can be selected with \fI+proj\fP\&. \fBproj \-l=id\fP
gives expanded description of projection \fIid\fP, e.g. \fBproj \-l=merc\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-lp
List of all projection id that can be used with the \fI+proj\fP parameter.
Equivalent to \fBproj \-l\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-lP
Expanded description of all projections that can be used with the \fI+proj\fP
parameter.
.UNINDENT
.INDENT 0.0
.TP
.B \-le
List of all ellipsoids that can be selected with the \fI+ellps\fP parameters.
.UNINDENT
.INDENT 0.0
.TP
.B \-lu
List of all distance units that can be selected with the \fI+units\fP parameter.
.UNINDENT
.INDENT 0.0
.TP
.B \-r
This options reverses the order of the expected input from
longitude\-latitude or x\-y to latitude\-longitude or y\-x.
.UNINDENT
.INDENT 0.0
.TP
.B \-s
This options reverses the order of the output from x\-y or longitude\-latitude
to y\-x or latitude\-longitude.
.UNINDENT
.INDENT 0.0
.TP
.B \-S
Causes estimation of meridional and parallel scale factors, area scale
factor and angular distortion, and maximum and minimum scale factors to be
listed between <> for each input point. For conformal projections meridional
and parallel scales factors will be equal and angular distortion zero. Equal
area projections will have an area factor of 1.
.UNINDENT
.INDENT 0.0
.TP
.B \-m <mult>
The cartesian data may be scaled by the \fImult\fP parameter. When processing data
in a forward projection mode the cartesian output values are multiplied by
\fImult\fP otherwise the input cartesian values are divided by \fImult\fP before inverse
projection. If the first two characters of \fImult\fP are 1/ or 1: then the
reciprocal value of \fImult\fP is employed.
.UNINDENT
.INDENT 0.0
.TP
.B \-f <format>
Where \fIformat\fP is a printf format string to control the form of the output values.
For inverse projections, the output will be in degrees when this option is
employed. The default format is \fB\(dq%.2f\(dq\fP for forward projection and DMS for
inverse.
.UNINDENT
.INDENT 0.0
.TP
.B \-w<n>
Where \fIn\fP is the number of significant fractional digits to employ for seconds
output (when the option is not specified, \fB\-w3\fP is assumed).
.UNINDENT
.INDENT 0.0
.TP
.B \-W<n>
Where \fIn\fP is the number of significant fractional digits to employ for seconds
output. When \fB\-W\fP is employed the fields will be constant width
with leading zeroes.
.UNINDENT
.INDENT 0.0
.TP
.B \-v
Causes a listing of cartographic control parameters tested for and used by
the program to be printed prior to input data.
.UNINDENT
.INDENT 0.0
.TP
.B \-V
This option causes an expanded annotated listing of the characteristics of
the projected point. \fI\%\-v\fP is implied with this option.
.UNINDENT
.sp
The \fI+opt\fP run\-line arguments are associated with cartographic parameters.
Additional projection control parameters may be contained in two auxiliary
control files: the first is optionally referenced with the
\fI+init=file:id\fP and the second is always processed after the name of the
projection has been established from either the run\-line or the contents of
+init file. The environment parameter \fI\%PROJ_DATA\fP establishes the
default directory for a file reference without an absolute path. This is
also used for supporting files like datum shift files.
.sp
Added in version 9.3.0: \fI{crs}\fP is one of the possibilities accepted by :c:\fBproj_create()\fP, provided it
expresses a projected CRS, like a WKT string, an object code (like \(dqEPSG:32632\(dq)
a PROJJSON string, etc.
The projection computed will be those of the map projection implied by
the transformation from the base geographic CRS of the projected CRS to the projected CRS.

.sp
One or more files (processed in left to right order) specify the source of
data to be converted. A \fB\-\fP will specify the location of processing standard
input. If no files are specified, the input is assumed to be from stdin.
For ASCII input data the two data values must be in the first two white space
separated fields and when both input and output are ASCII all trailing
portions of the input line are appended to the output line.
.sp
Input geographic data (longitude and latitude) must be in DMS or decimal degrees format and input
cartesian data must be in units consistent with the ellipsoid major axis or
sphere radius units. Output geographic coordinates will be in DMS (if the
\fB\-w\fP switch is not employed) and precise to 0.001\(dq with trailing, zero\-valued
minute\-second fields deleted.
.SH EXAMPLE
.sp
The following script
.INDENT 0.0
.INDENT 3.5
.sp
.EX
proj +proj=utm +zone=12 \-r <<EOF
45d15\(aq33.1\(dq 111.5W
45d15.551666667N \-111d30
+45.25919444444 111d30\(aq000w
EOF
.EE
.UNINDENT
.UNINDENT
.sp
will perform UTM zone 12 forward projection.
The default +ellps=GRS80 is used as no +ellps was specified.
The geographic values of this example are equivalent
and meant as examples of various forms of DMS input. The x\-y output
data will appear as three lines of:
.INDENT 0.0
.INDENT 3.5
.sp
.EX
460770.43     5011865.86
.EE
.UNINDENT
.UNINDENT
.sp
This other example
.INDENT 0.0
.INDENT 3.5
.sp
.EX
proj EPSG:6421 \-V <<EOF
\-120 35.8
EOF
.EE
.UNINDENT
.UNINDENT
.sp
Will perform the projection of the coordinates in \(dqNAD83(2011) / California zone 4\(dq
(\fIEPSG:6421\fP) into its geographic system, \(dqNAD83(2011)\(dq, showing the expanded annotated listing.
The output will appear as:
.INDENT 0.0
.INDENT 3.5
.sp
.EX
#Lambert Conformal Conic
#   Conic, Sph&Ell
#   lat_1= and lat_2= or lat_0, k_0=
# +proj=lcc +lat_0=35.3333333333333 +lon_0=\-119 +lat_1=37.25 +lat_2=36
# +x_0=2000000 +y_0=500000 +ellps=GRS80
#Final Earth figure: ellipsoid
#  Major axis (a): 6378137.000
#  1/flattening: 298.257222
#  squared eccentricity: 0.006694380023
Longitude: 120dW [ \-120 ]
Latitude:  35d48\(aqN [ 35.8 ]
Easting (x):   1909606.87
Northing (y):  552253.58
Meridian scale (h) : 1.00004382  ( 0.004382 % error )
Parallel scale (k) : 1.00004382  ( 0.004382 % error )
Areal scale (s):     1.00008765  ( 0.008765 % error )
Angular distortion (w): 0.000
Meridian/Parallel angle: 90.00000
Convergence : \-0d35\(aq47.714\(dq [ \-0.59658715 ]
Max\-min (Tissot axis a\-b) scale error: 1.00004 1.00004
.EE
.UNINDENT
.UNINDENT
.SH OTHER PROGRAMS
.sp

NEWS
9.4.1 Release Notes
-------------------

 Updates
 -------

 o Database: update to EPSG v11.006 (#4086)

 o Database: update ESRI dataset to ArcGIS Pro 3.3 (#4143)

 Bug fixes
 -------

 o DatabaseContext::lookForGridInfo(): do not trigger network activity in
   PROJ_GRID_AVAILABILITY_KNOWN_AVAILABLE mode when all grids are known (#4090)

 o Retry "Connection reset by peer" (#4107)

 o Use unique names for NKG transformations (#4111)

 o proj_create()/proj_create_from_database(): recall CRS AUTH:CODE in error message (#4114)

 o DerivedProjectedCRS: deal with lack of explicit CS in BASEPROJCRS in WKT (#4139)

 o proj_symbol_rename.h: update with current symbols (#4145)

 o Improve detection of compatible/incompatible celestial bodies (#4150)


9.4.0 Release Notes
-------------------

 Updates
 -------

 o Add support for transformations involving coordinate epoch changes
   (PointMotionOperation), specifically for Canadian NAD83(CSRS)(#3884)

 o SSL_OPTIONS: set SSL options on Curl library (#3936)

 o Add support for unity builds (#3962)

 o Added ability to install *.tif if present in data (#3970)

 o createOperationsCompoundToGeog(): tune selection logic
   when --grid-check known_available is specified (#3990)

 o Increase CMake minimum version from 3.9 to 3.16 (#3997)

 o CMake: use FetchContent to get googletest-1.12.1 (#4006)

 o CMake: Replace custom FindSqlite3 with FindSQLite3 built-in (#4007)

 o tpeqd: use numerically stable formula for computing the central angle
   from (phi_1, lam_1) to (phi_2, lam_2) (#4009)

 o Move content of proj_experimental.h to proj.h (#4019)

 o Add +proj=mod_krovak projection method for Modified Krovak that
   applies to S-JTSK/05 in the Czech Republic (#4021)

 o PROJString formatter optimizer: simplify pipelines doing
   [Modified]Krovak (South West) <--> [Modified]Krovak (East North)
   by just doing an axis swap (#4034)

 o +proj=gridshift: enhance to support grids referenced in projected CRS,
   and with easting_offset/northing_offset corrections (#4038)

 o Tune concatenated operation instanciation, reference CR2005 geoid for
   Czechia and add (disabled by default) records for Czechia S-JTSK/05 based
   transformations (#4044)

 o Database: update to EPSG v11.004 (#4062)

 Bug fixes
 ---------

 o Fix missing symbols at link time for Windows target in Visual Studio (#3984)

 o Improve error message in axisswap (#3975)

 o Avoid convergence errors in +proj=gridshift when using biquadratic
    interpolation (#3985)


9.3.1 Release Notes
-------------------

 Updates
 -------

 o Update to EPSG 10.098 (#3968)

 o Update ESRI objects to v3.2.0 (#3944)

 Bug fixes
 ---------

 o ITRF2008: fix wrong sign for 'dry' parameter of EURA and EURA_T  (#3870)

 o Fix build error with MSVC 2019 in /std:c++20 on NN_NO_CHECK() (#3872)

 o  ESRI WKT import: normalize GCS_unknown to unknown and D_unknown to unknown (#3874)

 o CoordinateOperationFactory: deal with CompoundToCompound with a horizontal similarity
   transformation and a ballpark vertical (#3881)

 o Ellipsoid::_isEquivalentTo(): fix so that an ellipsoid of semi-major axis A (and
   non-zero inv flattening) isn't equivalent to a sphere of radius A (#3882)

 o isEquivalentTo(): make a datum name 'unknown' equivalent to another one (#3883)

 o cs2cs: fix handling of input coordinates in grad (#3886)

 o Make setargv.obj available on Universal Windows Platform (#3891)

 o Allow opening proj.db with a URI (#3892)

 o createOperations(): fix GeogCRS 3D with TOWGS84 to geocentric CRS (#3915)

 o Fix test suite so that it can pass with ENABLE_TIFF=OFF (#3916)

 o GeographicBoundingBox::intersects(): avoid infinite recursion and stack overflow
   on invalid bounding boxes (#3919)

 o Fix importing '+proj=topocentric ... +type=crs' by using a geocentric CRS as
   the base CRS (#3924)

 o Allow LOCAL_CS with 3 axes (#3928)

 o WKT1 parser: in non-strict mode, accept missing UNIT[] in GEOGCS, GEOCCS,
   PROJCS and VERT_CS elements (#3933)

 o createOperations(): fix issue with a obscure case involving CompoundCRS of
   unknown horizontal datum + boundCRS of vertical (#3934)

 o createOperations(): fix bad PROJ pipeline when converting between Geog3D
   with non-metre height to CompoundCRS (#3943)

 o createOperations(): Fix possible null dereference on invalid WKT input (#3946)

 o proj_factor: fix when input is a compound CRS of a projected CRS (#3950)

 o pj_get_suggested_operation(): tune it to give correct result for RGAF09 to
   RRAF 1991 / UTM zone 20N + Guadeloupe 1988 height transformation (#3954)


src/proj.cpp
/* <<<< Cartographic projection filter program >>>> */
#include "proj.h"
#include "projects.h"
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <math.h>
#include "emess.h"

#if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(__WIN32__)
#  include <fcntl.h>
#  include <io.h>
#  define SET_BINARY_MODE(file) _setmode(_fileno(file), O_BINARY)
#else
#  define SET_BINARY_MODE(file)
#endif

#define MAX_LINE 1000
#define MAX_PARGS 100
#define PJ_INVERS(P) (P->inv ? 1 : 0)

extern void gen_cheb(int, projUV(*)(projUV), char *, PJ *, int, char **);

static PJ *Proj;
static union {
    projUV (*generic)(projUV, PJ *);
    projXY (*fwd)(projLP, PJ *);
    projLP (*inv)(projXY, PJ *);
} proj;

static int
    reversein = 0,  /* != 0 reverse input arguments */
    reverseout = 0, /* != 0 reverse output arguments */
    bin_in = 0,     /* != 0 then binary input */
    bin_out = 0,    /* != 0 then binary output */
    echoin = 0,     /* echo input data to output line */
    tag = '#',      /* beginning of line tag character */
    inverse = 0,    /* != 0 then inverse projection */
    prescale = 0,   /* != 0 apply cartesian scale factor */
    dofactors = 0,  /* determine scale factors */
    very_verby = 0, /* very verbose mode */
    postscale = 0;

static char
    *cheby_str,         /* string controlling Chebychev evaluation */
    *oform = (char *)nullptr, /* output format for x-y or decimal degrees */
    oform_buffer[16];   /* Buffer for oform when using -d */

static const char
    *oterr = "*\t*",    /* output line for unprojectable input */
    *usage = "%s\nusage: %s [ -bdeEfiIlmorsStTvVwW [args] ] [ +opts[=arg] ] [ files ]\n";

static PJ_FACTORS facs;

static double (*informat)(const char *, char **), /* input data deformatter function */
              fscale = 0.;                        /* cartesian scale factor */

static projUV int_proj(projUV data) {
    if (prescale) {
        data.u *= fscale;
        data.v *= fscale;
    }

    data = (*proj.generic)(data, Proj);

    if (postscale && data.u != HUGE_VAL) {
        data.u *= fscale;
        data.v *= fscale;
    }

    return data;
}

/* file processing function */
static void process(FILE *fid) {
    char line[MAX_LINE+3], *s = nullptr, pline[40];
    PJ_COORD data;

    for (;;) {
        int facs_bad = 0;
        ++emess_dat.File_line;

        if (bin_in) {   /* binary input */
            if (fread(&data, sizeof(projUV), 1, fid) != 1)
                break;
        } else {    /* ascii input */
            if (!(s = fgets(line, MAX_LINE, fid)))
                break;

            if (!strchr(s, '\n')) { /* overlong line */
                int c;
                (void)strcat(s, "\n");
                /* gobble up to newline */
                while ((c = fgetc(fid)) != EOF && c != '\n') ;
            }

            if (*s == tag) {
                if (!bin_out)
                    (void)fputs(line, stdout);
                continue;
            }

            if (reversein) {
                data.uv.v = (*informat)(s, &s);
                data.uv.u = (*informat)(s, &s);
            } else {
                data.uv.u = (*informat)(s, &s);
                data.uv.v = (*informat)(s, &s);
            }

            if (data.uv.v == HUGE_VAL)
                data.uv.u = HUGE_VAL;

            if (!*s && (s > line)) --s; /* assumed we gobbled \n */
            if (!bin_out && echoin) {
                char t;
                t = *s;
                *s = '\0';
                (void)fputs(line, stdout);
                *s = t;
                putchar('\t');
            }
        }

        if (data.uv.u != HUGE_VAL) {
            PJ_COORD coord;
            coord.lp = data.lp;
            if (prescale) { data.uv.u *= fscale; data.uv.v *= fscale; }
            if (dofactors && !inverse) {
                facs = proj_factors(Proj, coord);
                facs_bad = proj_errno(Proj);
            }

            data.xy = (*proj.fwd)(data.lp, Proj);

            if (dofactors && inverse) {
                facs = proj_factors(Proj, coord);
                facs_bad = proj_errno(Proj);
            }

            if (postscale && data.uv.u != HUGE_VAL)
                { data.uv.u *= fscale; data.uv.v *= fscale; }
        }

        if (bin_out) { /* binary output */
            (void)fwrite(&data, sizeof(projUV), 1, stdout);
            continue;
        } else if (data.uv.u == HUGE_VAL) /* error output */
            (void)fputs(oterr, stdout);
        else if (inverse && !oform) {   /*ascii DMS output */
            if (reverseout) {
                (void)fputs(rtodms(pline, data.uv.v, 'N', 'S'), stdout);
                putchar('\t');
                (void)fputs(rtodms(pline, data.uv.u, 'E', 'W'), stdout);
            } else {
                (void)fputs(rtodms(pline, data.uv.u, 'E', 'W'), stdout);
                putchar('\t');
                (void)fputs(rtodms(pline, data.uv.v, 'N', 'S'), stdout);
            }
        } else {    /* x-y or decimal degree ascii output, scale if warranted by output units */
            if (inverse) {
                if (proj_angular_input(Proj, PJ_FWD)) {
                    data.uv.v *= RAD_TO_DEG;
                    data.uv.u *= RAD_TO_DEG;
                }
            } else {
                if (proj_angular_output(Proj, PJ_FWD)) {
                    data.uv.v *= RAD_TO_DEG;
                    data.uv.u *= RAD_TO_DEG;
                }
            }

            if (reverseout) {
                (void)printf(oform, data.uv.v); putchar('\t');
                (void)printf(oform, data.uv.u);
            } else {
                (void)printf(oform, data.uv.u); putchar('\t');
                (void)printf(oform, data.uv.v);
            }
        }

        /* print scale factor data */
        if (dofactors) {
            if (!facs_bad)
                (void)printf("\t<%g %g %g %g %g %g>",
                    facs.meridional_scale, facs.parallel_scale, facs.areal_scale,
                    facs.angular_distortion * RAD_TO_DEG, facs.tissot_semimajor, facs.tissot_semiminor);
            else
                (void)fputs("\t<* * * * * *>", stdout);
        }
        (void)fputs(bin_in ? "\n" : s, stdout);
    }
}

/* file processing function --- verbosely */
static void vprocess(FILE *fid) {
    char line[MAX_LINE+3], *s, pline[40];
    LP dat_ll;
    projXY dat_xy;
    int linvers;
    PJ_COORD coord;


    if (!oform)
        oform = "%.3f";

    if (bin_in || bin_out)
        emess(1,"binary I/O not available in -V option");

    for (;;) {
        proj_errno_reset(Proj);
        ++emess_dat.File_line;

        if (!(s = fgets(line, MAX_LINE, fid)))
            break;

        if (!strchr(s, '\n')) { /* overlong line */
            int c;
            (void)strcat(s, "\n");
            /* gobble up to newline */
            while ((c = fgetc(fid)) != EOF && c != '\n') ;
        }

        if (*s == tag) { /* pass on data */
            (void)fputs(s, stdout);
            continue;
        }

        /* check to override default input mode */
        if (*s == 'I' || *s == 'i') {
            linvers = 1;
            ++s;
        } else
            linvers = inverse;

        if (linvers) {
            if (!PJ_INVERS(Proj)) {
                emess(-1,"inverse for this projection not avail.\n");
                continue;
            }
            dat_xy.x = strtod(s, &s);
            dat_xy.y = strtod(s, &s);
            if (dat_xy.x == HUGE_VAL || dat_xy.y == HUGE_VAL) {
                emess(-1,"lon-lat input conversion failure\n");
                continue;
            }
            if (prescale) { dat_xy.x *= fscale; dat_xy.y *= fscale; }
            if (reversein) {
                projXY temp = dat_xy;
                dat_xy.x = temp.y;
                dat_xy.y = temp.x;
            }
            dat_ll = pj_inv(dat_xy, Proj);
        } else {
            dat_ll.lam = proj_dmstor(s, &s);
            dat_ll.phi = proj_dmstor(s, &s);
            if (dat_ll.lam == HUGE_VAL || dat_ll.phi == HUGE_VAL) {
                emess(-1,"lon-lat input conversion failure\n");
                continue;
            }
            if (reversein) {
                LP temp = dat_ll;
                dat_ll.lam = temp.phi;
                dat_ll.phi = temp.lam;
            }
            dat_xy = pj_fwd(dat_ll, Proj);
            if (postscale) { dat_xy.x *= fscale; dat_xy.y *= fscale; }
        }

        /* For some reason pj_errno does not work as expected in some   */
        /* versions of Visual Studio, so using pj_get_errno_ref instead */
        if (*pj_get_errno_ref()) {
            emess(-1, pj_strerrno(*pj_get_errno_ref()));
            continue;
        }

        if (!*s && (s > line)) --s; /* assumed we gobbled \n */
        coord.lp = dat_ll;
        facs = proj_factors(Proj, coord);
        if (proj_errno(Proj)) {
            emess(-1,"failed to compute factors\n\n");
            continue;
        }

        if (*s != '\n')
            (void)fputs(s, stdout);

        (void)fputs("Longitude: ", stdout);
        (void)fputs(proj_rtodms(pline, dat_ll.lam, 'E', 'W'), stdout);
        (void)printf(" [ %.11g ]\n", dat_ll.lam * RAD_TO_DEG);
        (void)fputs("Latitude:  ", stdout);
        (void)fputs(proj_rtodms(pline, dat_ll.phi, 'N', 'S'), stdout);
        (void)printf(" [ %.11g ]\n", dat_ll.phi * RAD_TO_DEG);
        (void)fputs("Easting (x):   ", stdout);
        (void)printf(oform, dat_xy.x); putchar('\n');
        (void)fputs("Northing (y):  ", stdout);
        (void)printf(oform, dat_xy.y); putchar('\n');
        (void)printf("Meridian scale (h) : %.8f  ( %.4g %% error )\n", facs.meridional_scale, (facs.meridional_scale-1.)*100.);
        (void)printf("Parallel scale (k) : %.8f  ( %.4g %% error )\n", facs.parallel_scale, (facs.parallel_scale-1.)*100.);
        (void)printf("Areal scale (s):     %.8f  ( %.4g %% error )\n", facs.areal_scale, (facs.areal_scale-1.)*100.);
        (void)printf("Angular distortion (w): %.3f\n", facs.angular_distortion * RAD_TO_DEG);
        (void)printf("Meridian/Parallel angle: %.5f\n", facs.meridian_parallel_angle * RAD_TO_DEG);
        (void)printf("Convergence : ");
        (void)fputs(proj_rtodms(pline, facs.meridian_convergence, 0, 0), stdout);
        (void)printf(" [ %.8f ]\n", facs.meridian_convergence * RAD_TO_DEG);
        (void)printf("Max-min (Tissot axis a-b) scale error: %.5f %.5f\n\n", facs.tissot_semimajor, facs.tissot_semiminor);
    }
}

int main(int argc, char **argv) {
    char *arg, **eargv = argv, *pargv[MAX_PARGS], **iargv = argv;
    FILE *fid;
    int pargc = 0, iargc = argc, eargc = 0, mon = 0;

    if ( (emess_dat.Prog_name = strrchr(*argv,DIR_CHAR)) != nullptr)
        ++emess_dat.Prog_name;
    else emess_dat.Prog_name = *argv;
    inverse = ! strncmp(emess_dat.Prog_name, "inv", 3);
    if (argc <= 1 ) {
        (void)fprintf(stderr, usage, pj_get_release(), emess_dat.Prog_name);
        exit (0);
    }

    /* process run line arguments */
    while (--argc > 0) { /* collect run line arguments */
        if(**++argv == '-') for(arg = *argv;;) {
            switch(*++arg) {
              case '\0': /* position of "stdin" */
                if (arg[-1] == '-') eargv[eargc++] = "-";
                break;
              case 'b': /* binary I/O */
                bin_in = bin_out = 1;
                continue;
              case 'v': /* monitor dump of initialization */
                mon = 1;
                continue;
              case 'i': /* input binary */
                bin_in = 1;
                continue;
              case 'o': /* output binary */
                bin_out = 1;
                continue;
              case 'I': /* alt. method to spec inverse */
                inverse = 1;
                continue;
              case 'E': /* echo ascii input to ascii output */
                echoin = 1;
                continue;
              case 'V': /* very verbose processing mode */
                very_verby = 1;
                mon = 1;
                continue;
              case 'S': /* compute scale factors */
                dofactors = 1;
                continue;
              case 't': /* set col. one char */
                if (arg[1]) tag = *++arg;
                else emess(1,"missing -t col. 1 tag");
                continue;
              case 'l': /* list projections, ellipses or units */
                if (!arg[1] || arg[1] == 'p' || arg[1] == 'P') {
                    /* list projections */
                    const struct PJ_LIST *lp;
                    int do_long = arg[1] == 'P', c;
                    const char *str;

                    for (lp = proj_list_operations() ; lp->id ; ++lp) {
                        if( strcmp(lp->id,"latlong") == 0
                            || strcmp(lp->id,"longlat") == 0
                            || strcmp(lp->id,"geocent") == 0 )
                            continue;

                        (void)printf("%s : ", lp->id);
                        if (do_long)  /* possibly multiline description */
                            (void)puts(*lp->descr);
                        else { /* first line, only */
                            str = *lp->descr;
                            while ((c = *str++) && c != '\n')
                                putchar(c);
                            putchar('\n');
                        }
                    }
                } else if (arg[1] == '=') { /* list projection 'descr' */
                    const struct PJ_LIST *lp;

                    arg += 2;
                    for (lp = proj_list_operations(); lp->id ; ++lp)
                        if (!strcmp(lp->id, arg)) {
                            (void)printf("%9s : %s\n", lp->id, *lp->descr);
                            break;
                        }
                } else if (arg[1] == 'e') { /* list ellipses */
                    const struct PJ_ELLPS *le;

                    for (le = proj_list_ellps(); le->id ; ++le)
                        (void)printf("%9s %-16s %-16s %s\n",
                                     le->id, le->major, le->ell, le->name);
                } else if (arg[1] == 'u') { /* list units */
                    const struct PJ_UNITS *lu;

                    for (lu = proj_list_units(); lu->id ; ++lu)
                        (void)printf("%12s %-20s %s\n",
                                     lu->id, lu->to_meter, lu->name);
                } else if (arg[1] == 'd') { /* list datums */
                    const struct PJ_DATUMS *ld;

                    printf("__datum_id__ __ellipse___ __definition/comments______________________________\n" );
                    for (ld = pj_get_datums_ref(); ld->id ; ++ld)
                    {
                        printf("%12s %-12s %-30s\n",
                               ld->id, ld->ellipse_id, ld->defn);
                        if( ld->comments != nullptr && strlen(ld->comments) > 0 )
                            printf( "%25s %s\n", " ", ld->comments );
                    }
                } else
                    emess(1,"invalid list option: l%c",arg[1]);
                exit(0);
                /* cppcheck-suppress duplicateBreak */
                continue; /* artificial */
              case 'e': /* error line alternative */
                if (--argc <= 0)
                    noargument:
                emess(1,"missing argument for -%c",*arg);
                oterr = *++argv;
                continue;
              case 'T': /* generate Chebyshev coefficients */
                if (--argc <= 0) goto noargument;
                cheby_str = *++argv;
                continue;
              case 'm': /* cartesian multiplier */
                if (--argc <= 0) goto noargument;
                postscale = 1;
                if (!strncmp("1/",*++argv,2) ||
                    !strncmp("1:",*argv,2)) {
                    if((fscale = atof((*argv)+2)) == 0.)
                        goto badscale;
                    fscale = 1. / fscale;
                } else
                    if ((fscale = atof(*argv)) == 0.) {
                      badscale:
                        emess(1,"invalid scale argument");
                    }
                continue;
              case 'W': /* specify seconds precision */
              case 'w': /* -W for constant field width */
              {
                int c = arg[1];
                if (c != 0 && isdigit(c)) {
                    set_rtodms(c - '0', *arg == 'W');
                    ++arg;
                } else
                    emess(1,"-W argument missing or non-digit");
                continue;
              }
              case 'f': /* alternate output format degrees or xy */
                if (--argc <= 0) goto noargument;
                oform = *++argv;
                continue;
              case 'd':
                if (--argc <= 0) goto noargument;
                sprintf(oform_buffer, "%%.%df", atoi(*++argv));
                oform = oform_buffer;
                break;
              case 'r': /* reverse input */
                reversein = 1;
                continue;
              case 's': /* reverse output */
                reverseout = 1;
                continue;
              default:
                emess(1, "invalid option: -%c",*arg);
                break;
            }
            break;
        } else if (**argv == '+') { /* + argument */
            if (pargc < MAX_PARGS)
                pargv[pargc++] = *argv + 1;
            else
                emess(1,"overflowed + argument table");
        } else /* assumed to be input file name(s) */
            eargv[eargc++] = *argv;
    }
    if (eargc == 0 && !cheby_str) /* if no specific files force sysin */
        eargv[eargc++] = "-";
    else if (eargc > 0 && cheby_str) /* warning */
        emess(4, "data files when generating Chebychev prohibited");
    /* done with parameter and control input */
    if (inverse && postscale) {
        prescale = 1;
        postscale = 0;
        fscale = 1./fscale;
    }
    if (!(Proj = pj_init(pargc, pargv)))
        emess(3,"projection initialization failure\ncause: %s",
              pj_strerrno(pj_errno));

    if (!proj_angular_input(Proj, PJ_FWD)) {
        emess(3, "can't initialize operations that take non-angular input coordinates");
        exit(0);
    }

    if (proj_angular_output(Proj, PJ_FWD)) {
        emess(3, "can't initialize operations that produce angular output coordinates");
        exit(0);
    }

    if (inverse) {
        if (!Proj->inv)
            emess(3,"inverse projection not available");
        proj.inv = pj_inv;
    } else
        proj.fwd = pj_fwd;
    if (cheby_str) {
        gen_cheb(inverse, int_proj, cheby_str, Proj, iargc, iargv);
        exit(0);
    }

    /* set input formatting control */
    if (mon) {
        pj_pr_list(Proj);
        if (very_verby) {
            (void)printf("#Final Earth figure: ");
            if (Proj->es != 0.0) {
                (void)printf("ellipsoid\n#  Major axis (a): ");
                (void)printf(oform ? oform : "%.3f", Proj->a);
                (void)printf("\n#  1/flattening: %.6f\n",
                             1./(1. - sqrt(1. - Proj->es)));
                (void)printf("#  squared eccentricity: %.12f\n", Proj->es);
            } else {
                (void)printf("sphere\n#  Radius: ");
                (void)printf(oform ? oform : "%.3f", Proj->a);
                (void)putchar('\n');
            }
        }
    }

    if (inverse)
        informat = strtod;
    else {
        informat = proj_dmstor;
        if (!oform)
            oform = "%.2f";
    }

    if (bin_out) {
        SET_BINARY_MODE(stdout);
    }

    /* process input file list */
    for ( ; eargc-- ; ++eargv) {
        if (**eargv == '-') {
            fid = stdin;
            emess_dat.File_name = "<stdin>";

            if (bin_in)
            {
                SET_BINARY_MODE(stdin);
            }

        } else {
            if ((fid = fopen(*eargv, "rb")) == nullptr) {
                emess(-2, *eargv, "input file");
                continue;
            }
            emess_dat.File_name = *eargv;
        }
        emess_dat.File_line = 0;
        if (very_verby)
            vprocess(fid);
        else
            process(fid);
        (void)fclose(fid);
        emess_dat.File_name = nullptr;
    }

    if( Proj )
        pj_free(Proj);

    exit(0); /* normal completion */
}
